github/repos_releases.go

```diff
@@ -266,33 +266,40 @@ func (s *RepositoriesService) GetReleaseAsset(ctx context.Context, owner, repo s
 // If a redirect is returned, the redirect URL will be returned as a string instead
 // of the io.ReadCloser. Exactly one of rc and redirectURL will be zero.
 //
+// If followRedirects is true, a single redirect will be followed.
+//
 // GitHub API docs: https://developer.github.com/v3/repos/releases/#get-a-single-release-asset
-func (s *RepositoriesService) DownloadReleaseAsset(ctx context.Context, owner, repo string, id int64) (rc io.ReadCloser, redirectURL string, err error) {
+func (s *RepositoriesService) DownloadReleaseAsset(ctx context.Context, owner, repo string, id int64, followRedirects bool) (rc io.ReadCloser, redirectURL string, err error) {
 	u := fmt.Sprintf("repos/%s/%s/releases/assets/%d", owner, repo, id)
+	return s.downloadReleaseAssetFromURL(ctx, u, followRedirects)
+}
 
-	req, err := s.client.NewRequest("GET", u, nil)
+func (s *RepositoriesService) downloadReleaseAssetFromURL(ctx context.Context, url string, followRedirects bool) (rc io.ReadCloser, redirectURL string, err error) {
+	req, err := s.client.NewRequest("GET", url, nil)
 	if err != nil {
 		return nil, "", err
 	}
 	req.Header.Set("Accept", defaultMediaType)
 
 	s.client.clientMu.Lock()
-	defer s.client.clientMu.Unlock()
-
 	var loc string
 	saveRedirect := s.client.client.CheckRedirect
 	s.client.client.CheckRedirect = func(req *http.Request, via []*http.Request) error {
 		loc = req.URL.String()
 		return errors.New("disable redirect")
 	}
 	defer func() { s.client.client.CheckRedirect = saveRedirect }()
-
 	req = withContext(ctx, req)
 	resp, err := s.client.client.Do(req)
+	s.client.clientMu.Unlock()
```

Comment:
This mutex is no longer protecting `s.client.client.CheckRedirect` in line 291 such that this function is no longer re-entrant.

By this I mean that it is no longer safe to call this method from multiple threads, as the lock will be released here on line 294 in one thread, the next thread will get the lock and modify `s.client.client.CheckRedirect`, then the first thread will stomp all over `s.client.client.CheckRedirect` again in line 291 when it exits.

As it was written before, the first thread completely grabs the lock, and the defer on line 291 restores the protected variable, then the defer on line 280 is executed *after* that to release the lock, and finally the second thread can come in and do its thing (in the same manner).

I think we are going to have to come up with a different solution here, as I don't think we can use this one.

Reply:
@gmlewis I completely missed the `defer`. 
I added a new commit to restore `s.client.client.CheckRedirect` directly before `clientMu.Unlock()`.
 I assume that the CheckRedirect has returned when `s.client.client.Do(req)` has returned.
Will this not work? Am I missing something.

Reply:
Ah yes, very nice.  I believe this will work, @andersjanmyr.
Can you please try it out and verify it is working as desired?

Reply:
@gmlewis It's not working! Please hold off on the merge. I'll come back with a working version soon.

Reply:
No problem, @andersjanmyr!

Thank you very much for catching this before we check it in!  Your effort is greatly appreciated!